### ISSUE TITLE
Potential class generic issue: Type 'TypeName' [1] is incompatible with 'TypeName' [2]

### ISSUE THREAD

**Issue by Ricardo-Marques** [facebook/flow#6362]
Here is the shortest example I could get going with [Try Flow](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgCqAdgCZx4BOAPACoAWAlsQOYB8YAvGABQCUXDgDc4jUujwAPHHEoYwpPFACGAVxjyAxjGUBnXWAAiFSgGEd+uk1YcA3qgCQ5LrwzWWALjANmLPl6IyEytfDk47RwdKPAxVSmIwN0ZdADoWGJJyKh4k1j5HAF90B3SMTJMXHPcvHzyvcqoQm0FeAXCwWyKCoA)

```javascript
type Undoer<Thing> = () => void

export default class DoerClass<Thing> {
	do = (thing: Thing):  Undoer<Thing> => {
		return this.getUndoer(thing)
	}

	getUndoer = (thing: Thing): Undoer<Thing> => () => {}
}
```

Error is
> Cannot return `this.getUndoer(...)` because `Thing` [1] is incompatible with `Thing` [2] in type argument `Thing` [3]

**Comment by chisui** [comment]
Those seem to be illegal ways to define methods. If I strip all type annotations the code doesn't load in node. It stops as `do = ` and complains about the `=`.

If you use the standard method syntax your example [works in flow](https://flow.org/try/#0PTACDMBsHsHcCgAuBPADgUwAQFUB2ATadAJwB4BBAPkwF5MAKASluoDdoBLfeedAD1TRiiTPnTgAhgFdIIgMaQJAZyWYAIkWIBhRStIAhagG94mUdHqIAFh1wBzAFyZ9jJ3kIkDx02czF0iFLEuJjWHEoAdHYB7pqWNvaMPgC+PtGIsSTxto7OrjgEml6YJr5+AUEhTCwlqWapyUA)

**Comment by Ricardo-Marques** [comment]
Oh interesting, so it's a bug with generic types applied to class transformed properties. Thank you for the find!

**Comment by villesau** [comment]
I have similar, if not exactly the same problem: https://flow.org/try/#0FAegVABAAgZgNgewO4TCYBLAtgBwQJwBdUIBDAZwgCUBTUgY2JnwSwgHJ87H2BuYYIQCeOGhAAKLHOQA8AFQAWGAHYBzAHwQAvBADewCBATKAwgtJqaALggAKQkrU3FK1QEptmgG4IMAEwAaAwgHV2dHVWAAXwFhUQgXNTkRMR1dUhsARgAmAGYY4HpjcmIAZQBXACMTbTscKXIbXVCnBIjk0QCjU3NLG3sI8NcOmg8tb18-KLHNGT8MLxB1ARoADzwiCD8aGFJyuGJ6OApKABEEGnwTY-JZRI0INcIaZT9KWgZCADoTVjxlF6EGSSBDSeQRdSafSGLivS62DzQwwQLiEcr4ZR2GQVarBZEhCJaZpKchfeqg0ktVQxfGGYxmCyqGhE0Kk8nSL703pMmnIpZuYIxKJAA


My generic is not compatible with my type.

**Comment by chisui** [comment]
@villesau Your types actually don't match: `SubC` has the type `(props: Props<ThingType>) => ...` but in your `Com` class `props` has the type `Props<Thing>` where `Thing` is a type variable. So the errors in line 17 and 18 are actual errors. In line 17 you try to assign a `Thing` to `ThingType` and in line 18 a `(thing: Thing) => void` to `(thing: ThingType) => void`.

Depending on what you are trying to do you could change the code like [this](https://flow.org/try/#0JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwBQdMAnmFnAArGuxMA8AKgD44AXjgBvOnDgQAdgGEAFslkBzLAC44ACgBuyADYBXTXH4BKUcN0RgAEwA0UuDEXA1W-nQC+DAPR+cACSBHBMEEZwAO4q8DAQcADWWFhgZnDqslhQwLh0uHKo8ADKRgBG8qJwAoLaYFxanBDczDWWIsLavHbAun6C5vmF8PIkVTV1DRxc2a1C7Z28pRUSAHTr9c3e-YPSe3t0AcGh4ZExsnEJZe526YwsbMUkWE0tTFXiKFoAjABMAMxwXxDWRFOBPEBYZaVMSTZqNGY8JbPV6zJgDKw6bq9HYgsEQrCjUiwzZgBHNNHIyGongYjpY6FrDZcbYDIA).

**Comment by villesau** [comment]
@chisui I don't follow. The `Thing` needs to be generic. Subcomponent should be able to define it's type. This is simplified version where `Com` would get `SubCom` as a parameter. So when I pass component taking e.g object `{a: number}` to `Com`, `Com` should be able to require that from the parent. I'm aware that `Thing` is type variable - and it should not have opinion about the actual type.

So in a nutshell: `Com` needs to be generic without opinion about `thing`, but subComponent has an opinion about the props it can handle.

**Comment by chisui** [comment]
Ok lets go by your [reduced example](https://flow.org/try/#0FAegVABAAgZgNgewO4TCYBLAtgBwQJwBdUIBDAZwgCUBTUgY2JnwSwgHJ87H2BuYYIQCeOGhAAKLHOQA8AFQAWGAHYBzAHwQAvBADewCBATKAwgtJqaALggAKQkrU3FK1QEptmgG4IMAEwAaYABfAWFRCBc1ORExHV1SGwBGACYAZlDgemNyYgBlAFcAIxNtOxwpchtdYzMLVWs7B1dnR1UY0Q8tb18-YK7NGT8MLxB1ARoADzwiCD8aGFICuGJ6OApKE1Z5Ns0pwhplP0paBkIAOi3cY0PCGUkEaR3XdU19Qy4jmnxbD3fDCBcQgFfDKOwyQolAwAwy1cyWLS6Zrkc4VR4ouH1GihGFjNzQ0LBIA).

In line 11 you define the props of `SubC` as `{onChange: (thing: ThingType) => void}` which is equivalent to `Props<ThingType>`.

In line 13 you define `Com` as a component that has props of type `Props<Thing>`. The `Thing` part here is not a concrete type but a type variable. This variable is only bound when you use `Com`.

In line 16 you pass the `onChange` property of `Com`s props to `SubC`. For this assignment to typecheck both have to have the same type. The type of `this.props.onChange` is `(thing: Thing) => void` where `Thing` is the type variable of `Com`. The type of `onChange` that `SubC` expects is `(thing: ThingType) => void`. These two types are incompatible and that is essentially what the compile error says.

flow works as expected in your case, so this issue has nothing to do with the bug discussed in this issue.

**Comment by villesau** [comment]
@chisui Sorry for sidetracking! I think my examples took this into wrong direction: I just reduced wrongly which resulted similar result. This is what I experience in my current project [try flow](https://flow.org/try/#0FAegVABAAgZgNgewO4TCYBLAtgBwQJwBdUIBDAZwgCUBTUgY2JnwSwgHJ87H2BuYYIQCeOGhADKhGjgA8AFQAWGAHYBzAHwQAvBADeAYVYAuat0IA6Q7gTKaywnJE0ZuwkrUnFK1QBoIN-QVSNRoTAAo3b093VQBKbU0ANwQMABMAX3V0gWFRCAAFFhxyeRjNHV1gCAhyKWKTABJaUlSAeWU4IQBBfHxSIRlJaVLvdXUfKv9lQODVUIgImOjveK0klNSJ6siPCC81LZqAVwAjADUMGiQTZsZLVjxbe0dRFx3VZYOpmZDw98+4gkIMk0plgNlBE49jEXmIKqQTABGABMAGYIfQbLUJKd9NoFjgiuQTK4ltDvLC-AEgr8Fv9yWpYat1qDmRAZKkMIkQOpgJjlNjlEcsCYhVgTjR8PiUaj+DQAB54IgQVI0GCkI5wYj0OAUShWEZqTQKqTKVKUW4WKyPOyEGSFBDFQ0aTSVaqY6xPQgAMQIQxw+LCbMi5HMhMdodq0nIAG0xQBde5YfjVLhmyVBvSTd1Y4jiXGI-Eh8wem32X34f2ZkAgCCEfBCEz87H5k76Qs6Yvh4rmKPFOPCxNWFPVCBcQhHfDKBaDAvZ0fvLSkjCh7uh97ZUej6mzGhLrtE8w7kKb0c82KTbLZIA)

Note the commented out part. If you try with that instead of function call, Flow succeeds. Though, in this case changing it to standard syntax does not help, only getting rid of function call altogether helps.

**Comment by chisui** [comment]
@villesau Yeah, now it seems to be the same issue.

**Comment by chisui** [comment]
Interestingly it also helps to define a explicit type annotation. [try flow](https://flow.org/try/#0MYGwhgzhAECCA8AVAfNA3gKGtAZtAvNABQAuAXNIgJQGoAMW0A5gceZTfqiQBYCWEAHQ5SVDAF8MGUJBgAhJKkzYcFIqQrVa0OpzabO9Ri0IaO23gOGiJQA).

It looks like flow simply fails to infer the type of `A.f` in that example. `flow type-at-pos` claims that it has the type `any`.

**Comment by Ricardo-Marques** [comment]
Seems like the example above with the inline typed definitions for each method is also broken with the current release.

[Here is another interesting replication of this bug, relating to interface `implement` with generics](https://flow.org/try/#0PTACDMBsHsHcCgCWA7ALgUwE7gIYGN0ACASQEFlEBbHSAHgBUBPAB3QD5CBveQwgagDm6VE1YAuQgAoAlIQC8HUengBfePDyQcAZ22FyVGgxbtCVZpHSV0aPWQrU6Sjt14B9VCYlKehX3mhkbVRMAFc8VGhMSU9xQiVZV15CVAALRG0AOg8TeRSTXzVfIRETGS5fXkxhUMxkFPSsnNZC1SA)

**Comment by SamChou19815** [comment]
This now works.
